### ISSUE TITLE
Dialog: close button, optional title, layout

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/joist#413]
@ariel-phet @emily-phet can we add "Close" buttons to the "About" and "Check for Updates" dialogs?

For a11y, all modal dialogs should have a button that can receive focus when the dialog is opened, and so that the dialog can be dismissed without needing to press "escape".

The close button should not be optional for dialogs, but that part should likely be addressed with #359.

**Comment by emily-phet** [comment]
@jessegreenberg We should be consistent with the standards here. If the standards say a close button is needed, we should have this. No reason to get too creative for a simple dialog!

**Comment by ariel-phet** [comment]
@jessegreenberg I see no issue with that, agreed that it can just be added as a requirement to #359

**Comment by jessegreenberg** [comment]
Thanks @emily-phet and @ariel-phet, I will add a close button to these dialogs.

**Comment by phet-steele** [comment]
@jessegreenberg (all) PhET-iO sims are failing automated testing with this button, here is an example error from CAF:

```
Error: Assertion failed: Cannot register two different instances to the same id: chargesAndFields.navigationBar.phetButton.phetMenu.aboutDialog.closeButton
    at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/assert/js/assert.js:21:13)
    at Object.addInstance (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/phet-io/js/phetio.js?bust=1491334162126:413:19)
    at Object.addInstance (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/phet-io/js/phetio.js?bust=1491334162126:57:14)
    at Tandem.addInstance (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/tandem/js/Tandem.js?bust=1491334162126:126:36)
    at RectangularPushButton.setTandem (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/scenery/js/nodes/Node.js?bust=1491334162126:3467:22)
    at RectangularPushButton.set tandem [as tandem] (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/scenery/js/nodes/Node.js?bust=1491334162126:3472:32)
    at https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/scenery/js/nodes/Node.js?bust=1491334162126:4763:25
    at u (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/sherpa/lib/lodash-4.17.4.min.js?bust=1491334162126:5:432)
    at Function.ru (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/sherpa/lib/lodash-4.17.4.min.js?bust=1491334162126:67:656)
    at RectangularPushButton.mutate (https://bayes.colorado.edu/continuous-testing/snapshot-1491334002353/scenery/js/nodes/Node.js?bust=1491334162126:4750:9)
Approximately 4/4/2017, 1:26:42 PM
```

**Comment by jessegreenberg** [comment]
Thanks @phet-steele. In the above commit, we remove the tandem instance from the close button when disposed, phet-io tests are passing.

**Comment by jessegreenberg** [comment]
Ok, NOW it is fixed.

**Comment by pixelzoom** [comment]
@jessegreenberg Where you aware of the problems with joist.Dialog close button? See https://github.com/phetsims/joist/issues/346.

Since you've added the close button to the About dialog, it now overlaps the title for some sims, with `?brand=adapted-from-phet` (the default).  See screenshot below.  Perhaps this isn't a big deal since we don't publish `adapted-from-phet`.   But I wanted you to be aware of this problem - these these changes with a bunch of sims.

![screenshot_230](https://cloud.githubusercontent.com/assets/3046552/25391625/e30adc26-2993-11e7-855a-d35c9a290734.png)

**Comment by jessegreenberg** [comment]
Thanks @pixelzoom, I will take a look at this.

**Comment by pixelzoom** [comment]
@jessegreenberg @ariel-phet If a close button is indeed necessary for a11y, then someone might want to take a look at all PhET sims, to see how many have dialogs without close buttons.  By default, `joist.Dialog` does _not_ have a close button (`hasCloseButton: false`).  And adding a close button in some sims won't be as easy as just changing `hasCloseButton`.   Some sims (e.g. unit-rates) don't use `joist.Dialog` because of problems like #346 and the lack of progress on fixing Dialog (#359).

**Comment by pixelzoom** [comment]
This is going to be complicated by the fact that some dialogs (eg, KeyboardHelpDialog) create their own custom close button, and add it independently of Dialog.  When we make close button required, KeyboardHelpDialog will have 2 close buttons!

**Comment by pixelzoom** [comment]
molecules-and-light/SpectrumWindowDialog has a "Close" button at the bottom center of the dialog. It will need to be removed.

**Comment by andrea-phet** [comment]
Sims that specify a value for hasCloseButton:
curve-fitting
- [ ] ReducedChiSquaredStatisticDialog
equality-explorer
- [ ] OopsDialog
- [ ] InfoDialog
gene-expression-essentials
- [ ] FluorescentCellsPictureDialog
joist
- [ ] AboutDialog
- [ ] DialogsScreenView
- [ ] KeyboardHelpDialog
- [ ] OptionsDialog
- [ ] UpdateDialog
make-a-ten
- [ ] InfoDialog
models-of-the-hydrogen-atom
- [ ] SnapshotsDialog
molecules-and-light
- [ ] SpectrumWindowDialog
scenery-phet
- [ ] ContextLossFailureDialog
vegas
- [ ] RewardDialog

**Comment by zepumph** [comment]
I'm not sure if this is the best place for this, but I have found the recent ui update to close button breaking keyboard interaction. The reason is that `RectangularPushButton` had accessibleInputListeners (search `addAccessibleInputListener`), but `Path` doesn't. If we want to keep it as a path, then we will need to implement these listeners ourselves. 

If going down this road, my recommendation is to change the `closeButtonListener` option back to a function, rather than a `ButtonListener` (whose "down" event is only useful for mouse interaction), and then add the function to the "click" event for the accessible listener. Let me know if I can be of more help here.

**Comment by pixelzoom** [comment]
@andrea-phet let's discuss @zepumph's comment above.

**Comment by pixelzoom** [comment]
@andrea-phet let's discuss @zepumph's comment above.  In order to support a11y and phet-io, I think it would be best if the close button were a `RectangularPushButton` with transparent `baseColor`.

**Comment by pixelzoom** [comment]
@andrea-phet let's discuss @zepumph's comment above.  In order to support a11y and phet-io, I think it would be best if the close button were a `RectangularPushButton` with `baseColor: 'transparent'`.  

@zepumph would that solve the problem?

**Comment by zepumph** [comment]
I committed a temporary fix, but it isn't over writable or customizable with options.

**Comment by zepumph** [comment]
Ahh, sorry @pixelzoom I didn't see any of your comments until just now. Yeah, I think it would be best to style it, but still have it be a `RectangularPushButton`.

**Comment by pixelzoom** [comment]
I'm going to work on an immediate fix.

**Comment by pixelzoom** [comment]
Looks like this also got removed from the revised close button:

```js
      accessibleFire: function() {
        self.focusActiveElement();
      },
```

@zepumph What does this do, and are there problems without it?

**Comment by pixelzoom** [comment]
Two dialogs have problems with the close button being requires:

(1) KeyboardHelpDialog creates its own custom close button, exactly on top of the standard close button, to be addressed in https://github.com/phetsims/scenery-phet/issues/374.

(2) ContextLossFailureDialog has overlap of content and button, to be addressed in https://github.com/phetsims/scenery-phet/issues/373.

**Comment by pixelzoom** [comment]
I cleaned up all of the client usages related to the close button.

We need to review the positioning of the close button.  As currently implemented in `updateCloseButtonPosition`, it's possible for the close button to overlap the content (as I saw in ContextLossFailureDialog).  I'll have to review the design document to see what the specification says.

**Comment by pixelzoom** [comment]
As specified, the design in https://docs.google.com/document/d/12UkhG7lU-dLBCa-A8FcamkrPBLQxr6ft33Igbw6a5U8/edit#heading=h.wwa6byadzd8h allows the close button to overlap the content.  But I doubt that's intention -- the specification appears to indicate a lack of understanding about how layout, margins, etc. work.  So we'll need to discuss with @amanda-phet and @ariel-phet.

**Comment by andrea-phet** [comment]
@pixelzoom could you review my updates to the layout? Also, is https://github.com/phetsims/joist/issues/477 addressed by the design changes?

**Comment by pixelzoom** [comment]
Notes/sketch from 5/7/18 meeting with @andrea-phet, to refresh my memory on the "requirements".

![dialog-design-spec](https://user-images.githubusercontent.com/3046552/40750244-e11ac5fe-6423-11e8-8436-bab72cb5c683.jpg)

**Comment by pixelzoom** [comment]
Changes look good.    But now that all dialogs have a title area, we'll need to alert developers.  Because some of the dialog look odd now.  For example, see the About and Update dialogs below.   We'll need to decide whether to add titles to these, decide how to adjust margins, etc.   I'll add developer-meeting label to discuss at tomorrow's meeting.

![screenshot_656](https://user-images.githubusercontent.com/3046552/40750567-f4c02742-6424-11e8-954f-b4764168eff0.png)

![screenshot_657](https://user-images.githubusercontent.com/3046552/40750571-f809612a-6424-11e8-9ce6-8fa094e181c9.png)

**Comment by pixelzoom** [comment]
@andrea-phet We should also look at the default for some of the options.   E.g., these seem a bit extreme:

```js
      xSpacing: 20, // {number} how far the title is placed to the left of the close button
      ySpacing: 20, // {number} how far the title is placed above the content

      xMargin: 20,
      yMargin: 20,
```

**Comment by andrea-phet** [comment]
@pixelzoom agreed.

**Comment by pixelzoom** [comment]
I'm starting to question the decision to have a title area (possibly blank) for all Dialogs.  I have the Dialog shown below in Equality Explorer.  I'm not convinced that it needs a title, and it looks lopsided with a blank area at the top. 

![screenshot_658](https://user-images.githubusercontent.com/3046552/40759629-29fd643a-6450-11e8-8674-a678826a5ec0.png)

**Comment by amanda-phet** [comment]
Do all dialogs *need* a title? 

In the sun demo example, sun demo is obviously the title. But we definitely need the option to not have a title, or we will need to do custom spacing for all dialogs that do not have a title.

**Comment by jonathanolson** [comment]
> It would be better to rethink the layout for title-less Dialogs and have it handled consistently by common code.

That sounds good to me. Presumably we'll still want to provide an accessible "name" to all dialogs.

**Comment by pixelzoom** [comment]
I've sketched out 3 design options for Dialogs that don't have a title, see below. (And feel free to suggest others.)

Option 1 - Close button overlaps dialog content. Up to the developer to ensure that they don't overlap. Could be tricky with i18n.

Option 2 - Close button to the right of content. Equal margin on left and right of content.

Option 3 - Close button above content. Equal margin above and below content.

![screenshot_659](https://user-images.githubusercontent.com/3046552/40795454-6196f6e4-64bf-11e8-97c4-97ef01d8e4f3.png)

**Comment by pixelzoom** [comment]
Of the above 3, my preference would be Option 3.   While Option 2 & 3 are similar, Option 3 is a better fit for the aspect ratio of ScreenView layoutBounds.

**Comment by pixelzoom** [comment]
Of the above 3, my preference would be Option 2.   While Option 2 & 3 are similar, Option 2 is a better fit for the aspect ratio of ScreenView layoutBounds.

**Comment by pixelzoom** [comment]
Option 1 is something that I really think we should strive to avoid.   And I'd also like to avoid implementing multiple layout options solely for the sake of individual aesthetic preferences.  

With Option 2, I don't think a bit of extra horizontal spacing would look bad.  Especially if you had the option to set the xSpacing between the close button and content to zero.  It seems like a small price to pay for the alternative of having to deal with implementing multiple layouts and deal with potential overlap in client code.

**Comment by pixelzoom** [comment]
About an Option 4, like Option 2 but with different margins on the left and right?  For content that is left justified (e.g. About dialog) perhaps this would eliminate the "excess horizontal spacing" concern.

**Comment by pixelzoom** [comment]
5/31/18 design meeting:

We'll go with a version of Option 2 for both titled and untitled dialogs, with an option to change the left margin.  I will sketch out the new design, and discuss with @andrea-phet.

**Comment by samreid** [comment]
At today's meeting, I suggested we look at the iOS dialog layout to solve symmetry problems:

![image](https://user-images.githubusercontent.com/679486/40802566-2d65f938-64d3-11e8-8de5-f08957c2b0ed.png)

We decided to continue investigating the "x" though (I'm not 100% clear why though).

**Comment by andrea-phet** [comment]
Dialog sketch from @pixelzoom on 5/31:

![img_9216](https://user-images.githubusercontent.com/7988696/40995472-61b62aa4-68bc-11e8-925c-0f9f2cf3071f.jpg)

**Comment by andrea-phet** [comment]
@pixelzoom, could you review Dialog? I think my next step if current changes are good would be to go into use cases and fix any workarounds. One such example would be `ContextLossFailureDialog` in joist:
```js
Dialog.call( this, new HBox( {
  children: [ warningSign, text, button ],
  spacing: 10
} ), {
  // margins large enough to make space for close button, see https://github.com/phetsims/scenery-phet/issues/373
  xMargin: 50,
  yMargin: 30
} );
```

**Comment by pixelzoom** [comment]
This looks overall great.   

The only thing I see is potential confusion about options `leftMargin/rightMargin/topMargin/bottomMargin` vs `xMargin/yMargin`.  I see that the former are applied to the content, while the latter are applied to the Panel (supertype).   Unless you have other reasons for wanting to provide both sets of options, I think it might be preferable to get rid of  `xMargin` and `yMargin`, and keep the Panel's margins set to zero like this:

```js
assert && assert( options.xMargin === undefined, 'Dialog sets xMargin' );
options.xMargin = 0;
assert && assert( options.yMargin === undefined, 'Dialog sets yMargin' );
options.yMargin = 0;
```

And yes, a good next step is to review/adjust usages of Dialog.

**Comment by andrea-phet** [comment]
Current uses of Dialog:

curve-fitting
- [ ] ReducedChiSquaredStatisticDialog

equality-explorer
- [ ] InfoDialog
- [ ] LevelSelectionNode
- [ ] OopsDialog

gene-expression-essentials
- [ ] FluorescentCellsPictureDialog

joist
- [ ] AboutDialog
- [ ] KeyboardHelpDialog
- [ ] OptionsDialog
- [ ] UpdateDialog

make-a-ten
- [ ] InfoDialog

masses-and-springs
- [ ] EnergyGraphNode

models-of-the-hydrogen-atom
- [ ] SnapshotsDialog

molecules-and-light
- [ ] SpectrumWindowDialog

pendulum-lab
- [ ] EnergyLegendDialog

plinko-probability
- [ ] OutOfBallsDialog

scenery-phet
- [ ] ContextLossFailureDialog

sun
- [ ] DialogsScreenView

vegas
- [ ] RewardDialog

**Comment by pixelzoom** [comment]
@andrea-phet and I discussed a problem with the current layout options.  If you increase the margins  around the content (or title + content), then the close button starts to move away from the upper-right corner.   For example in Equality Explorer:

![screenshot_677](https://user-images.githubusercontent.com/3046552/41553355-c5ef6e18-72ee-11e8-9b4b-335798063453.png)

What we really want is this, where the close button remains snugged up into the upper-right corner:

![screenshot_677](https://user-images.githubusercontent.com/3046552/41553855-41d11288-72f0-11e8-86d8-99c233acc704.jpg)

Here's the revised layout and options that I sketched out.  The closeButton has its own top and right margin options, which we will hopefully keep constant for a consistent "look".

![doc](https://user-images.githubusercontent.com/3046552/41553865-4b440140-72f0-11e8-8c0e-e69b78ae1c7e.jpg)

Let's try these defaults for the options:

```js
xSpacing: 10, // {number} how far the title and content is placed to the left of the close button
ySpacing: 10 // {number} vertical space between title and content
topMargin: 15, // {number} margin above content, or above title if provided 
bottomMargin: // {number} margin below content
 // {number|null} margin to the left of the content.  If null, this is computed so that we have
// the same margins on the left and right of the content.
leftMargin: null, 
closeButtonTopMargin: 10, // {number} margin above the close button
closeButtonRightMargin: 10 // {number} margin to the right of the close button 
```

And if `leftMargin` is not provided, it's computation is:

```js
options.leftMargin = options.xSpacing + CLOSE_BUTTON_WIDTH + options.closeButtonRightMargin;
```

**Comment by pixelzoom** [comment]
@andrea-phet The above changes look great.  I played around with various layout options in the sun demo, and it all seems to be working as expected.  I think this is a nice improvement.   I especially like the diagram that you added to explain the layout options, very helpful.  

Next step would be to (re)review the usages listed in https://github.com/phetsims/joist/issues/413#issuecomment-397433509.  You might want to uncheck that list and start over.

**Comment by pixelzoom** [comment]
I reviewed all of the Dialog usages listed in https://github.com/phetsims/joist/issues/413#issuecomment-397433509.  I tweaked a few, but they generally looked good.

**Comment by pixelzoom** [comment]
@ariel-phet Everything looks good to me here, and I would be comfortable with closing this issue.  But if the resources are available, you might have an impartial 3rd party review the final API.   Since Dialog still has other issues that have not been address, the scope of the review should be limited to what was addressed in this issue -- close button, title, and layout options.

**Comment by jessegreenberg** [comment]
Discussed 6/21/18 developer meeting:

@jessegreenberg will review the changes.

**Comment by jessegreenberg** [comment]
I reviewed the changes regarding layout, close button, and title. I played with all of the options and everything worked as expected, even with extreme values.

The diagram added to document the layout options was extremely helpful, that was a great addition. Everything worked as documented. Based on comments in https://github.com/phetsims/joist/issues/413#issuecomment-398569022, closing this issue. Thanks very much @andrealin and @pixelzoom.
